common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -31,34 +36,551 @@
+                                }
+                                //Show an error dialog only if a dialog is still showing.
+                                if (mDialogHolder.isDialogShowing()) {
+                                    //Call Cancel Cba callback
+                                    mDialogHolder.onCancelCba();
+                                    //Show dialog informing users that they have unplugged their device too early.
+                                    mDialogHolder.showErrorDialog(R.string.smartcard_early_unplug_dialog_title, R.string.smartcard_early_unplug_dialog_message);
+                                    Logger.verbose(TAG, "YubiKey was disconnected while dialog was still displayed.");
+                                }
+                            }
+                        }
+                    });
+                }
+            }
+        });
     }
 
-    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    /**
+     * Called when a ClientCertRequest is received by the AzureActiveDirectoryWebViewClient.
+     * Prompts the user to choose a certificate to authenticate with based on whether or not a YubiKey is plugged in and has permission to be connected.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     * @return null in either case.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
     @Override
     public Void processChallenge(@NonNull final ClientCertRequest request) {
         //final String methodTag = TAG + ":processChallenge";
-        return handleOnDeviceCertAuth(request);
+        //If a smartcard device is connected, proceed with smartcard CBA.
+        synchronized (sDeviceLock) {
+            if (mDevice != null) {
+                handleSmartcardCertAuth(request);
+                return null;
+            }
+        }
+        //Else, proceed with user certificates stored on device.
+        handleOnDeviceCertAuth(request);
+        return null;
+    }
+
+    /**
+     * Collects YubiKitCertDetails from PIV certificates on YubiKey.
+     * If certificates are found on YubiKey, the method proceeds with the smartcard certificate picker.
+     * Otherwise, appropriate error dialogs are shown.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void handleSmartcardCertAuth(@NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + ":handleSmartcardCertAuth";
+        //A connection to the YubiKey needs to be made in order to read the certificates off it.
+        getActivePivSessionAsync(request, new IPivSessionCallback() {
+            @Override
+            public void onGetSession(@NonNull PivSession piv) {
+                try {
+                    //Check if there are no PIN attempts remaining.
+                    if (piv.getPinAttempts() == 0) {
```

Comment:
Not sure if you're saying it's redundant to check here or pointing out that the logic is repeated. 
I'll extract this logic into a method in a future PR (where I will be restructuring the desigin to have YubiKit calls in a separate class), but regarding the former, I wanted to check the pin attempts here because it seemed unnecessary for a user to even see the cert picker dialog if they were locked out of their YubiKey anyway.